File: 프로그래머스/0/120875. 평행/평행.cs
using System;
using System.Collections.Generic;

public class CartesianCoordinate
{
    public int X = 0;
    public int Y = 0;

    public bool IsSame(CartesianCoordinate other)
    {
        return X == other.X && Y == other.Y;
    }
}

public class LineInformation
{
    public CartesianCoordinate Point1 = new CartesianCoordinate();
    public CartesianCoordinate Point2 = new CartesianCoordinate();

    public double Slope
    {
        get
        {
            return (double)(Point2.Y - Point1.Y) / (Point2.X - Point1.X);
        }
    }

    public LineInformation(int x1, int y1, int x2, int y2)
    {
        Point1.X = x1;
        Point1.Y = y1;
        Point2.X = x2;
        Point2.Y = y2;
    }

    public bool IsPointShared(LineInformation other)
    {
        return Point1.IsSame(other.Point1) || Point1.IsSame(other.Point2) ||
            Point2.IsSame(other.Point1) || Point2.IsSame(other.Point2);
    }

    public bool IsParallel(LineInformation other)
    {
        return Slope == other.Slope;
    }
}

public class Solution
{
    public int solution(int[,] dots)
    {
        List<LineInformation> lines = new List<LineInformation>();

        for (int i = 0; i < dots.GetLength(0); i++)
        {
            for (int j = i + 1; j < dots.GetLength(0); j++)
            {
                var lineInfo = new LineInformation(dots[i, 0], dots[i, 1], dots[j, 0], dots[j, 1]);

                lines.Add(lineInfo);
            }
        }

        for (int i = 0; i < lines.Count; i++)
        {
            for (int j = i + 1; j < lines.Count; j++)
            {
                if (!lines[i].IsPointShared(lines[j]) && lines[i].IsParallel(lines[j]))
                {
                    return 1;
                }
            }
        }

        return 0;
    }
}

File: 프로그래머스/0/120875. 평행/평행.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

int solution(vector<vector<int>> dots) {
    // 각각 페어가 되는 경우의 수
    vector<vector<int>> pairs = {
        {0, 1, 2, 3},
        {0, 2, 1, 3},
        {0, 3, 1, 2}
    };

    for (vector<int> xy : pairs) {        
        int mul1 = (dots[xy[1]][1] - dots[xy[0]][1]) * (dots[xy[3]][0] - dots[xy[2]][0]);
        int mul2 = (dots[xy[1]][0] - dots[xy[0]][0]) * (dots[xy[3]][1] - dots[xy[2]][1]);

        if (mul1 == mul2) {
            return 1;
        }
    }

    return 0;
}